[CosmosDB] Adding pointInTimeRestore commands to the stable branch.

amisi01 opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description
Adding point-in-time restore commands to the stable branch.
* Adding the tests for pitr commandlets
* Adding the help for the new commandlets
* Adding the point in time restore commandlets
* Fixing static analysis issue
* Flattening responses and adding pipeline inputs
* Fixing the timestamp for deleted accounts restore
* Adding detailed descriptions
* Addressing PR suggestions
* Adding missed files for PR suggestions
* Updating change log
* Adding suppressions for expected static analyzer errors
* Adding beking change exceptions
* Fixing the psdl file to pickup ps1xml file
* Update ChangeLog.md

Polish change log
* Update ChangeLog.md
* Update ChangeLog.md

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/701
        -  Azure/azure-powershell-cmdlet-review-pr#953        
    - [x] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

amisi01 commented:
[Don'tMergePRYet] Waiting for new .Net SDK to be released. Will update the PR with new SDK version change (3.0.1) and will update test record files for that too. In the meantime, please review the code changes and provide your valuable feedback/suggestion. Thanks

BethanyZhou commented:
- Have you raise a design doc for this change? https://github.com/Azure/azure-powershell-cmdlet-review-pr
- Please add online version for each help files to resolve CI failure, for example:
```
-  Target      : Get-AzCosmosDBMongoDBRestorableResource.md
  Description : Online version in the header of the file is incorrect. The corresponding URL should be the following: 
                https://docs.microsoft.com/powershell/module/az.cosmosdb/get-azcosmosdbmongodbrestorableresource
```

amisi01 commented:
> https://docs.microsoft.com/powershell/module/az.cosmosdb/get-azcosmosdbmongodbrestorableresource

Hi @BethanyZhou ,
Yes, i have raised the cmdlet preview and review is accepted - https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/953

Updated help files.

Please take a look. Thanks

BethanyZhou commented:
Hi @amisi01 , looks like you only proposed one cmdlet in this design doc. While I saw many cmdlets changed in this PR. Could you add them all in your design doc?

amisi01 commented:
> Hi @amisi01 , looks like you only proposed one cmdlet in this design doc. While I saw many cmdlets changed in this PR. Could you add them all in your design doc?

Rest of the changes were already checked in to preview branch(https://github.com/Azure/azure-powershell/commit/fba5e7d003ac528a4107599bb2d29ec24e33d87d). 
Only the one api was added after that for which I shared the cmdlet design doc.

BethanyZhou commented on src/CosmosDB/CosmosDB/Restore/GetAzCosmosDBMongoDBRestorableDatabase.cs:
```diff
@@ -0,0 +1,53 @@
+    {
+        [Parameter(Mandatory = true, ParameterSetName = NameParameterSet, HelpMessage = Constants.LocationNameHelpMessage)]
+        [ValidateNotNullOrEmpty]
+        public string LocationName { get; set; }
```
LocationName -> Location

amisi01 replied:
We have two following two field which refers to different things as per current convention in code:
Location: array of string (like the set of locations where account need to be created)
LocationName: string (single location where we want to fetch that particular record.)

Help string is also per the above convention

BethanyZhou replied:
Why we aren't able to combine them as one?

amisi01 replied:
We can combine them as one but there are following two issue with that:
1. Get is only for one location. Allowing Get to take input string array might be bit confusing to customer and also we have to add checks to make sure and string array only contains 1 element in it.
2. Help message is different for Location:string[] vs LocationName:string.

amisi01 replied:
Updated name to Location

BethanyZhou commented on src/CosmosDB/CosmosDB/ChangeLog.md:
```diff
@@ -19,6 +19,20 @@
+    - Get-AzCosmosDBMongoDBRestorableDatabase, Get-AzCosmosDBMongoDBRestorableCollection, Get-AzCosmosDBMongoDBRestorableResource.
+  - Removed the support for throughput Migration. (Please use the stable version of Az.CosmosDB Powershell module if you intend to use this feature).
+  - Removed AnalyticalStorageTTL from SqlContainer datamodel (Please use the stable version of Az.CosmosDB Powershell module if you intend to use this feature).
+* Please note this is a preview version. The cmdlets introduced here may be changed or removed in a future release.
```
Hi if you add these cmdlets in stable release, breaking change is only allowed in major release.

amisi01 replied:
Should we increase the major version if we add the new cmdlets only? (no breaking change to existing cmdlets)

BethanyZhou replied:
If no breaking to existing cmdlets, we needn't increase major version. But if you announce the cmdlets will change in future, I recommend you to avoid adding them in stable version

amisi01 replied:
I have updated the message in update log.

BethanyZhou commented on src/CosmosDB/CosmosDB/CosmosDBAccount/NewAzCosmosDBDatabaseToRestore.cs:
```diff
@@ -0,0 +1,39 @@
+namespace Microsoft.Azure.Commands.CosmosDB
+{
+    [Cmdlet(VerbsCommon.New, ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "CosmosDBDatabaseToRestore"), OutputType(typeof(PSDatabaseToRestore))]
+    public class NewAzCosmosDBDatabaseToRestore : AzureCosmosDBCmdletBase
```
Said `New-AzCosmosDBRestore` in https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/701, while it's `New-AzCosmosDBDatabaseToRestore` here, please have a look

amisi01 replied:
yeah, we updated the verb for Restore. New-AzCosmosDBRestore => Restore-AzCosmosDBAccount
And, the api (New-AzCosmosDBDatabaseToRestore), only creates the list of databases/collections that we need to restore and this is passed as argument to "Restore-AzCosmosDBAccount" api.

BethanyZhou replied:
So what's New-AzCosmosDBRestore for here?

amisi01 replied:
We don't have 'New-AzCosmosDBRestore' in stable cmdlets. We have renamed this one from preview version to Restore-AzCosmosDBAccount.

BethanyZhou commented on src/CosmosDB/CosmosDB/Restore/GetAzCosmosDBMongoDBRestorableCollection.cs:
```diff
@@ -0,0 +1,61 @@
+    {
+        [Parameter(Mandatory = true, ParameterSetName = NameParameterSet, HelpMessage = Constants.LocationNameHelpMessage)]
+        [ValidateNotNullOrEmpty]
+        public string LocationName { get; set; }
```
Please replace LocationName with Location

amisi01 replied:
We have two following two field which refers to different things as per current convention in code:
Location: array of string (like the set of locations where account need to be created)
LocationName: string (single location where we want to fetch that particular record.)

Help string is also per the above convention

BethanyZhou replied:
Are we able to use string location for Get, string array for other operations?

amisi01 replied:
Updated name to Location

BethanyZhou commented on src/CosmosDB/CosmosDB/Restore/GetAzCosmosDBMongoDBRestorableCollection.cs:
```diff
@@ -0,0 +1,61 @@
+
+        [Parameter(Mandatory = true, ParameterSetName = NameParameterSet, HelpMessage = Constants.DatabaseResourceIdHelpMessage)]
+        [ValidateNotNullOrEmpty]
+        public string DatabaseRid { get; set; }
```
I recommend using `DatabaseId` directly.

amisi01 replied:
Actually this field refers Database Resource Id which is different than Database Id. Used DatabaseRid just to avoid confusion with DatabaseId field.

BethanyZhou replied:
Make sense. How about DatabaseRid -> DatabaseRId to follow Pascal naming?

amisi01 replied:
done

BethanyZhou commented on src/CosmosDB/CosmosDB/Restore/GetAzCosmosDBMongoDBRestorableCollection.cs:
```diff
@@ -0,0 +1,61 @@
+
+        [Parameter(Mandatory = true, ValueFromPipeline = true, ParameterSetName = ParentObjectParameterSet, HelpMessage = Constants.RestorableMongoDBDatabaseObjectHelpMessage)]
+        [ValidateNotNull]
+        public PSRestorableSqlDatabaseGetResult ParentObject { get; set; }
```
We usually use InputObject instead.

amisi01 replied:
Updated

BethanyZhou commented on src/CosmosDB/CosmosDB/Restore/GetAzCosmosDBMongoDBRestorableDatabase.cs:
```diff
@@ -0,0 +1,53 @@
+
+        [Parameter(Mandatory = true, ValueFromPipeline = true, ParameterSetName = ParentObjectParameterSet, HelpMessage = Constants.RestorableDatabaseAccountObjectHelpMessage)]
+        [ValidateNotNull]
+        public PSRestorableDatabaseAccountGetResult ParentObject { get; set; }
```
ParentObject ->InputObject

amisi01 replied:
updated

BethanyZhou commented on src/CosmosDB/CosmosDB/Restore/GetAzCosmosDBMongoDBRestorableResource.cs:
```diff
@@ -0,0 +1,61 @@
+    {
+        [Parameter(Mandatory = true, HelpMessage = Constants.LocationNameHelpMessage)]
+        [ValidateNotNullOrEmpty]
+        public string LocationName { get; set; }
```
Location

amisi01 replied:
We have two following two field which refers to different things as per current convention in code:
Location: array of string (like the set of locations where account need to be created)
LocationName: string (single location where we want to fetch that particular record.)

Help string is also per the above convention

BethanyZhou commented on src/CosmosDB/CosmosDB/Restore/GetAzCosmosDBMongoDBRestorableResource.cs:
```diff
@@ -0,0 +1,61 @@
+
+        [Parameter(Mandatory = true, ValueFromPipeline = true, ParameterSetName = ParentObjectParameterSet, HelpMessage = Constants.RestorableDatabaseAccountObjectHelpMessage)]
+        [ValidateNotNull]
+        public PSRestorableDatabaseAccountGetResult ParentObject { get; set; }
```
InputObject

amisi01 replied:
updated

BethanyZhou commented on src/CosmosDB/CosmosDB/Restore/GetAzCosmosDBRestorableDatabaseAccount.cs:
```diff
@@ -0,0 +1,78 @@
+    {
+        [Parameter(Mandatory = false, HelpMessage = Constants.LocationNameHelpMessage)]
+        [ValidateNotNullOrEmpty]
+        public string LocationName { get; set; }
```
Location

amisi01 replied:
We have two following two field which refers to different things as per current convention in code:
Location: array of string (like the set of locations where account need to be created)
LocationName: string (single location where we want to fetch that particular record.)

Help string is also per the above convention

BethanyZhou commented on src/CosmosDB/CosmosDB/Restore/GetAzCosmosDBSqlRestorableContainer.cs:
```diff
@@ -0,0 +1,61 @@
+    {
+        [Parameter(Mandatory = true, ParameterSetName = NameParameterSet, HelpMessage = Constants.LocationNameHelpMessage)]
+        [ValidateNotNullOrEmpty]
+        public string LocationName { get; set; }
```
Location

amisi01 replied:
We have two following two field which refers to different things as per current convention in code:
Location: array of string (like the set of locations where account need to be created)
LocationName: string (single location where we want to fetch that particular record.)

Help string is also per the above convention

BethanyZhou commented on src/CosmosDB/CosmosDB/Restore/GetAzCosmosDBSqlRestorableContainer.cs:
```diff
@@ -0,0 +1,61 @@
+
+        [Parameter(Mandatory = true, ParameterSetName = NameParameterSet, HelpMessage = Constants.DatabaseResourceIdHelpMessage)]
+        [ValidateNotNullOrEmpty]
+        public string DatabaseRid { get; set; }
```
DatabaseId

amisi01 replied:
Actually this field refers Database Resource Id which is different than Database Id. Used DatabaseRid just to avoid confusion with DatabaseId field.

amisi01 replied:
updated

BethanyZhou commented on src/CosmosDB/CosmosDB/Restore/GetAzCosmosDBSqlRestorableContainer.cs:
```diff
@@ -0,0 +1,61 @@
+
+        [Parameter(Mandatory = true, ValueFromPipeline = true, ParameterSetName = ParentObjectParameterSet, HelpMessage = Constants.RestorableSqlDatabaseObjectHelpMessage)]
+        [ValidateNotNull]
+        public PSRestorableSqlDatabaseGetResult ParentObject { get; set; }
```
InputObject

amisi01 replied:
updated

BethanyZhou commented on src/CosmosDB/CosmosDB/Restore/GetAzCosmosDBSqlRestorableDatabase.cs:
```diff
@@ -0,0 +1,53 @@
+    {
+        [Parameter(Mandatory = true, ParameterSetName = NameParameterSet, HelpMessage = Constants.LocationNameHelpMessage)]
+        [ValidateNotNullOrEmpty]
+        public string LocationName { get; set; }
```
Location

amisi01 replied:
We have two following two field which refers to different things as per current convention in code:
Location: array of string (like the set of locations where account need to be created)
LocationName: string (single location where we want to fetch that particular record.)

Help string is also per the above convention

amisi01 replied:
Updated name to Location

BethanyZhou commented on src/CosmosDB/CosmosDB/Restore/GetAzCosmosDBSqlRestorableDatabase.cs:
```diff
@@ -0,0 +1,53 @@
+
+        [Parameter(Mandatory = true, ValueFromPipeline = true, ParameterSetName = ParentObjectParameterSet, HelpMessage = Constants.RestorableDatabaseAccountObjectHelpMessage)]
+        [ValidateNotNull]
+        public PSRestorableDatabaseAccountGetResult ParentObject { get; set; }
```
InputObject

amisi01 replied:
updated

BethanyZhou commented on tools/StaticAnalysis/Exceptions/Az.CosmosDB/BreakingChangeIssues.csv:
```diff
@@ -0,0 +1,19 @@
+"AssemblyFileName","ClassName","Target","Severity","ProblemId","Description","Remediation"
+"Microsoft.Azure.PowerShell.Cmdlets.CosmosDB.dll","Microsoft.Azure.Commands.CosmosDB.InvokeAzCosmosDBCassandraKeyspaceThroughputMigration","Invoke-AzCosmosDBCassandraKeyspaceThroughputMigration","0","1000","The cmdlet 'Invoke-AzCosmosDBCassandraKeyspaceThroughputMigration' has been removed and no alias was found for the original cmdlet name.","Add the cmdlet 'Invoke-AzCosmosDBCassandraKeyspaceThr ...
```
why you remove these cmdlets?

amisi01 replied:
Removed.

BethanyZhou commented on tools/StaticAnalysis/Exceptions/Az.CosmosDB/BreakingChangeIssues.csv:
```diff
@@ -0,0 +1,19 @@
+"Microsoft.Azure.PowerShell.Cmdlets.CosmosDB.dll","Microsoft.Azure.Commands.CosmosDB.InvokeAzCosmosDBMongoDBDatabaseThroughputMigration","Invoke-AzCosmosDBMongoDBDatabaseThroughputMigration","0","1000","The cmdlet 'Invoke-AzCosmosDBMongoDBDatabaseThroughputMigration' has been removed and no alias was found for the original cmdlet name.","Add the cmdlet 'Invoke-AzCosmosDBMongoDBDatabaseThroughputM ...
+"Microsoft.Azure.PowerShell.Cmdlets.CosmosDB.dll","Microsoft.Azure.Commands.CosmosDB.InvokeAzCosmosDBSqlContainerThroughputMigration","Invoke-AzCosmosDBSqlContainerThroughputMigration","0","1000","The cmdlet 'Invoke-AzCosmosDBSqlContainerThroughputMigration' has been removed and no alias was found for the original cmdlet name.","Add the cmdlet 'Invoke-AzCosmosDBSqlContainerThroughputMigration' ba ...
+"Microsoft.Azure.PowerShell.Cmdlets.CosmosDB.dll","Microsoft.Azure.Commands.CosmosDB.InvokeAzCosmosDBSqlDatabaseThroughputMigration","Invoke-AzCosmosDBSqlDatabaseThroughputMigration","0","1000","The cmdlet 'Invoke-AzCosmosDBSqlDatabaseThroughputMigration' has been removed and no alias was found for the original cmdlet name.","Add the cmdlet 'Invoke-AzCosmosDBSqlDatabaseThroughputMigration' back t ...
+"Microsoft.Azure.PowerShell.Cmdlets.CosmosDB.dll","Microsoft.Azure.Commands.CosmosDB.NewAzCosmosDBSqlContainer","New-AzCosmosDBSqlContainer","0","2000","The cmdlet 'New-AzCosmosDBSqlContainer' no longer supports the parameter 'AnalyticalStorageTtl' and no alias was found for the original parameter name.","Add the parameter 'AnalyticalStorageTtl' back to the cmdlet 'New-AzCosmosDBSqlContainer', or ...
```
why you delete this parameter?

amisi01 replied:
Removed

BethanyZhou commented on tools/StaticAnalysis/Exceptions/Az.CosmosDB/SignatureIssues.csv:
```diff
@@ -35,4 +35,5 @@
-"Microsoft.Azure.PowerShell.Cmdlets.CosmosDB.dll","Microsoft.Azure.Commands.CosmosDB.UpdateAzCosmosDBAccount","Update-AzCosmosDBAccount","1","8410","Parameter NetworkAclBypass of cmdlet Update-AzCosmosDBAccount does not follow the enforced naming convention of using a singular noun for a parameter name.","Consider using a singular noun for the parameter name."
\ No newline at end of file
+"Microsoft.Azure.PowerShell.Cmdlets.CosmosDB.dll","Microsoft.Azure.Commands.CosmosDB.UpdateAzCosmosDBAccount","Update-AzCosmosDBAccount","1","8410","Parameter NetworkAclBypass of cmdlet Update-AzCosmosDBAccount does not follow the enforced naming convention of using a singular noun for a parameter name.","Consider using a singular noun for the parameter name."
+"Microsoft.Azure.PowerShell.Cmdlets.CosmosDB.dll","Microsoft.Azure.Commands.CosmosDB.NewAzCosmosDBDatabaseToRestore","New-AzCosmosDBDatabaseToRestore","1","8100","New-AzCosmosDBDatabaseToRestore Does not support ShouldProcess but the cmdlet verb New indicates that it should.","Determine if the cmdlet should implement ShouldProcess and if so determine if it should implement Force / ShouldContinue"
```
Please add set shouldprocess=true for NewAzCosmosDBDatabaseToRestore

amisi01 replied:
Removed the change. Actually, this cmd (New-AzCosmosDBDatabaseToRestore) doesn't do any processing. It just creates the list of databases/collections to restore which gets used by restore cmdlet and have shouldprocess=true

amisi01 replied:
I had to keep this line for fixing static analysis error. But for this api, we don't need shouldprocess=true for this api as this api isn't an action api. Its just a helper api for RestoreAzCosmosDBDatabase api.

BethanyZhou reviewed (changes requested):
Hi Seems like there are some mismatches between design doc and PR, please check them.

- Breaking changes are not allowed in our regular release, but they are allowed in preview release or our major release. Please remove them from this PR.
- Some markdown files only shorten parameter type, like Type: System.String->String, please revert them.

kavskalyan commented on src/CosmosDB/CosmosDB/CosmosDBAccount/NewAzCosmosDBAccount.cs:
```diff
@@ -208,9 +224,27 @@ public override void ExecuteCmdlet()
+                databaseAccountCreateUpdateParameters.BackupPolicy = backupPolicy.ToSDKModel();
+            }
+
             if (BackupIntervalInMinutes.HasValue || BackupRetentionIntervalInHours.HasValue)
```
```suggestion
            else if (BackupIntervalInMinutes.HasValue || BackupRetentionIntervalInHours.HasValue)
```

kavskalyan commented on src/CosmosDB/CosmosDB/SQL/GetAzCosmosDBSqlContainerBackupInformation.cs:
```diff
@@ -0,0 +1,80 @@
+            }
+            catch (CloudException)
+            {
+                throw;
```
do we want to catch exception if we are doing a blind throw?

amisi01 replied:
Improved catch statement to handle notfound case and rethrow other errors.

amisi01 commented:
> Hi Seems like there are some mismatches between design doc and PR, please check them.
> 
> * Breaking changes are not allowed in our regular release, but they are allowed in preview release or our major release. Please remove them from this PR.
> * Some markdown files only shorten parameter type, like Type: System.String->String, please revert them.

No breaking change in this PR. Updated the changelog to address the mismatches.
Removed the markdown files only with the changes for parameter name shortening.

BethanyZhou commented on src/CosmosDB/CosmosDB/ChangeLog.md:
```diff
@@ -19,6 +19,17 @@
+  - Introduced support for Point in time restore for accounts with continuous mode backup policy.
+  - Introduced support to update the backup interval and backup retention for accounts with periodic mode backup policy.
+  - Introduced support to list the restorable resources in a live database account.
+  - New-AzCosmosBAccount cmdlet is updated with new parameters: IsRestoreReqeust, RestoreSource, RestoreTimestampInUtc, DatabasesToRestore, BackupPolicy.
+  - Update-AzCosmosDBAccount cmdlet is updated with new parameters: BackupPolicy.
```
```suggestion
  - Added new parameters IsRestoreReqeust, RestoreSource, RestoreTimestampInUtc and DatabasesToRestore,BackupPolicy in `New-AzCosmosBAccount`
  - Added BackupPolicy in `Update-AzCosmosDBAccount`
```
Suggest to shorten release log as possible due to the length limit for release note
I remembered IsRestoreReqeust is changed, please check it.

amisi01 replied:
Shortened the message.

BethanyZhou commented on src/CosmosDB/CosmosDB/CosmosDB.generated.format.ps1xml:
```diff
@@ -69,5 +69,197 @@
+                <Label>Type</Label>
+              </ListItem>
+              <ListItem>
+                <PropertyName>_rid</PropertyName>
+                <Label>_rid</Label>
```
Looks like _rid is private property, is there any public alternative for _rid? please replace it.
BTW, label can be different with PropertyName, I suggest ResourceId if rid means `resource id`.

amisi01 replied:
Updated to ResourceId

BethanyZhou commented on src/CosmosDB/CosmosDB/Restore/GetAzCosmosDBSqlRestorableResource.cs:
```diff
@@ -0,0 +1,73 @@
+    {
+        [Parameter(Mandatory = true, HelpMessage = Constants.LocationNameHelpMessage)]
+        [ValidateNotNullOrEmpty]
+        public string LocationName { get; set; }
```
Look like you missed this one, please check it

amisi01 replied:
updated
